src/core/selinux-access.c

```diff
@@ -239,7 +239,10 @@ int mac_selinux_access_check_internal(
                         if (!enforce)
                                 return 0;
 
-                        return sd_bus_error_setf(error, SD_BUS_ERROR_ACCESS_DENIED, "Failed to get file context on %s.", path);
+                        return sd_bus_error_setf(error,
+                                                 r == -ENOENT ? SD_BUS_ERROR_FILE_NOT_FOUND : SD_BUS_ERROR_ACCESS_DENIED,
+                                                 "Failed to get file context on %s.",
+                                                 path);
```

Comment:
this is ugly: there can be so many reasons this fails, ENOENT is just one among many, and specially handling that one, over all the other ones is a mess.

let's maybe fix this properly: cache the label in the `Unit` object whenever we load the file, and then load it from the open fd, instead of doing a separate path-based lookup here.

Doing second level security checks by paths is racy as fuck, and revolting to be frank.

If you do it that way, the issue goes away entirely, and we don't have to special case error codes anymore.

Reply:
this was proposed in some other issue already. and would make things a ton nicer: we could then report the label as bus property and so on, and the label would follow our usual load cycle.

Reply:
That method was objected at https://github.com/systemd/systemd/pull/10023#issuecomment-650025249

Reply:
> this is ugly: there can be so many reasons this fails, ENOENT is just one among many, and specially handling that one, over all the other ones is a mess.

But which other SD_BUS_ERROR value would match an errno? I struggle to see any that could be returned by that call?

Reply:
can that libselinux api actually return any of those?

Reply:
it's just a wrapper around getxattr(), no? and I doubt they eat up those errors... (but then again, my hopes might be too high)

Reply:
It does get bubbled up - according to the manpage, getxattr can return E2BIG, ERANGE, ENODATA, ENOTSUP. We shouldn't worry about the first one here, ERANGE is handled by the library with a retry, and ENOTSUP seems fine to map it to ACCESS_DENIED. I'll add ENODATA to the NOT_FOUND mapping.

Reply:
> > That method was objected at [#10023 (comment)](https://github.com/systemd/systemd/pull/10023#issuecomment-650025249)
> 
> Well, I disagree with the reasoning. If selinux wants to be taken seriously as a security tech, then the project should learn that applying/reading labels always via file system paths is a racy mess, and undestand we don#t really do things anymore, but rather want to pin inodes while we operate on them, and then get a reasonably congruent view of the world.

I am not going to solve that here. If all of that happens at some point in the future, all the better - but right now there's a very real problem (can't tell between a permission denial and a file that's gone), and I need a fix for it, and it can't really wait for a hypothetical major redesign that might arrive down the line

Reply:
I am not sure about `ENODATA`.
According to getxattr(2) there are two causes for it:
  - The named attribute does not exist,
  -  or the process has no access to this attribute.

On a SELinux enabled system all files *should* have a security context (named attribute of `"security.selinux"`) either by creation or initial file system setup and it can't be removed. So a unit file not having a SELinux context is very unlikely and serious.

Also for the second reason `ENODATA` should not be mapped to `SD_BUS_ERROR_FILE_NOT_FOUND`.

Reply:
> or the process has no access to this attribute.

I'm not sure this could realistically happen to pid1 given the existing policy? Which means it can only be the first one - which might be worrying, but it's entirely correct to map as NOT_FOUND. It doesn't mean it gets ignored, it just means there's a meaningful error reported back, so that one can do the appropriate action.

Reply:
> > On a SELinux enabled system all files _should_ have a security context (named attribute of `"security.selinux"`) either by creation or initial file system setup and it can't be removed. So a unit file not having a SELinux context is very unlikely and serious.
> 
> What happens if i mount a foreign (and thus unlabelled) ext4 fs on an selinux-enabled host? do its inodes magically acquire labels?

AFAIK no, it would all stay unlabeled, it would require a restorecon on the volume to apply labels according to the current policy

Reply:
> What happens if i mount a foreign (and thus unlabelled) ext4 fs on an selinux-enabled host? do its inodes magically acquire labels?

Then the file will indeed have no label. (The kernel falls back for access checks to the file initial security identifier - commonly `unlabeled_t`.)
But is `SD_BUS_ERROR_FILE_NOT_FOUND` a more meaningful error than `SD_BUS_ERROR_ACCESS_DENIED` in this case, as the actual file exists?

(I am not strongly against this change, just asking.)